src/backend/distributed/operations/citus_tools.c

```diff
@@ -492,12 +523,48 @@ ExecuteCommandsAndStoreResults(StringInfo *nodeNameArray, int *nodePortArray,
  * the calling user's credentials. The function returns the query status
  * (success/failure), and query result. The query is expected to return a single
  * target containing zero or one rows.
+ *
+ * This function forces a new connection that is closed after the query is run.
  */
 static bool
 ExecuteRemoteQueryOrCommand(char *nodeName, uint32 nodePort, char *queryString,
 							StringInfo queryResultString)
 {
 	int connectionFlags = FORCE_NEW_CONNECTION;
+	return ExecuteRemoteQueryOrCommandInternal(nodeName, nodePort, queryString,
+											   queryResultString, connectionFlags);
+}
+
+
+/*
+ * ExecuteRemoteQueryOrCommand executes a query at specified remote node using
+ * the calling user's credentials. The function returns the query status
+ * (success/failure), and query result. The query is expected to return a single
+ * target containing zero or one rows.
+ *
+ * This funtion does not force a new connection, and reuses existing connections if there are any
+ */
+static bool
+ExecuteRemoteQueryOrCommandUsingExistingConnections(char *nodeName, uint32 nodePort,
```

Comment:
The name `ExistingConnections` might be confusing. If there are no eligible connections, this function would also establish a new connection. One option could be  add a new flag to `ExecuteRemoteQueryOrCommand`, such as `bool forceNewConnection` and that decides on the `connectionFlags`

Reply:
I ended up adding several parameters to `ExecuteRemoteQueryOrCommand`

- bool forceNewConnection
- const char *user
- const char *database